File: 프로그래머스/2/84512. 모음 사전/모음 사전.cpp
#include <string>
#include <vector>

using namespace std;
int count, answer;
char alphabet[5] = {'A', 'E', 'I', 'O', 'U'};

void dfs(string word, string next) {
    if (next == word) {
        answer = count;
        return;
    }

    if (next.length() >= 5) {
        return;
    }

    for (int i=0; i<5; i++) {
        count += 1;
        dfs(word, next + alphabet[i]);
    }
}

int solution(string word) {
    dfs(word, "");
    return answer;
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.py
"""
백트래킹!
"""

def solution(word):
    count = 0
    answer = 0
    
    def backtrack(cur_word):
        nonlocal count, answer
        
        if answer > 0:
            return
        
        if cur_word == word:
            answer = count
            return
        
        if len(cur_word) >= 5: 
            return 
        
        for a in ['A', 'E', 'I', 'O', 'U']:
            count += 1
            backtrack(cur_word + a)
    backtrack("")
    return answer
